source/fuzz/fuzzer_pass_replace_opphi_ids_from_dead_predecessors.cpp

```diff
@@ -0,0 +1,115 @@
+    : FuzzerPass(ir_context, transformation_context, fuzzer_context,
+                 transformations) {}
+
+FuzzerPassReplaceOpPhiIdsFromDeadPredecessors::
+    ~FuzzerPassReplaceOpPhiIdsFromDeadPredecessors() = default;
+
+void FuzzerPassReplaceOpPhiIdsFromDeadPredecessors::Apply() {
+  // Keep a vector of the transformations to apply.
+  std::vector<TransformationReplaceOpPhiIdFromDeadPredecessor> transformations;
+
+  // Loop through the blocks in the module.
+  for (auto& function : *GetIRContext()->module()) {
+    for (auto& block : function) {
+      // Only consider dead blocks.
+      if (!GetTransformationContext()->GetFactManager()->BlockIsDead(
+              block.id())) {
+        continue;
+      }
+
+      // Find all the uses of the label id of the block inside OpPhi
+      // instructions.
+      GetIRContext()->get_def_use_mgr()->ForEachUse(
+          block.id(), [this, &function, &block, &transformations](
+                          opt::Instruction* instruction, uint32_t) {
+            // Only consider OpPhi instructions.
+            if (instruction->opcode() != SpvOpPhi) {
+              return;
+            }
+
+            // Randomly decide whether to consider this use.
+            if (!GetFuzzerContext()->ChoosePercentage(
+                    GetFuzzerContext()
+                        ->GetChanceOfReplacingOpPhiIdFromDeadPredecessor())) {
+              return;
+            }
+
+            // Get the current id corresponding to the predecessor.
+            uint32_t current_id = 0;
+            for (uint32_t i = 1; i < instruction->NumInOperands(); i += 2) {
+              if (instruction->GetSingleWordInOperand(i) == block.id()) {
+                // The corresponding id is at the index of the block - 1.
+                current_id = instruction->GetSingleWordInOperand(i - 1);
+                break;
+              }
+            }
+
+            // Get the type id.
+            uint32_t type_id = instruction->type_id();
+
+            // If the list of candidates for this type was not already computed,
+            // compute it.
+            const auto& candidates = FindAvailableInstructions(
+                &function, &block, block.end(),
+                [type_id, current_id](opt::IRContext* /* unused */,
+                                      opt::Instruction* candidate) -> bool {
+                  // Only consider instructions with a result id, with the
+                  // right type and different from the currently-used one.
+                  return candidate->HasResultId() &&
+                         candidate->type_id() == type_id &&
+                         candidate->result_id() != current_id;
+                });
+
+            // If there is no possible replacement, we cannot apply any
+            // transformation.
+            if (candidates.empty()) {
+              return;
+            }
+
+            // Choose one of the candidates.
```

Comment:
I like this comment BTW.  Not that the line of code is particularly complex, but it's complex enough that you don't see at a glance what it does.